### PR TITLE
[tactics] Remove anonymous fix/cofix form.

### PR DESCRIPTION
**Pull request by ejgallego**
We remove the `fix N / cofix N` forms from the tactic language. This
way, these tactics don't depend anymore on the proof context, in
particular on the proof name, which seems like a fragile practice.

Apart from the concerns wrt maintenability of proof scripts, this also
helps making the "proof state" functional; as we don't have to
propagate the proof name to the tactic layer.

Depends on #7379.

### CODE REVIEW

**Review by herbelin — COMMENTED** [review]
Something is indeed unsatisfactory with `fix n`/`cofix n`. No strong opinion on what to do. Reusing the lemma name would make sense if `fix`/`cofix` is at the head of the proof but otherwise it looks a bit arbitrary.

I would say that "add naming information inside the proof monad" should be the first thing to do. Then deprecating `fix n`/`cofix n` could also be considered. I'm wondering about what others think.

About having `tactics` depending on `proofs` (where by `proofs` is meant proof management, not proof engine), this is a good remark. Maybe, my 2p, do we want them independent of each other?

**Comment by ejgallego** [comment]
> About having tactics depending on proofs (where by proofs is meant proof management, not proof engine), this is a good remark. Maybe, my 2p, do we want them independent of each other?

Indeed it seems that `proofs` contains two totally unrelated things as of today, the pre-8.5 proof-engine, and the `start_proof` and friends. Maybe a split would indeed help.

**Comment by ejgallego** [comment]
I'm not sure adding "naming" to the proof monad wouldn't be counter productive in the long term, I think it is much better to grab the `Proof.t` object in `coretactics.ml4` and push that info down if needed.

**Comment by ejgallego** [comment]
Anyways deprecation seems well withing range given the travis results, we are talking about a very small patch it seems.

**Comment by ejgallego** [comment]
So if we provide instead a `TACTIC FUNCTIONAL EXTEND`, we'd do like this:
```ocaml
TACTIC FUNCTIONAL EXTEND fix
  [ "fix" natural(n) ] -> [ fun st -> let id = name_of_proof st.proof in Tactics.fix id n ]
| [ "fix" ident(id) natural(n) ] -> [ Tactics.fix id n ]
END
```

**Comment by ppedrot** [comment]
> I'm not sure adding "naming" to the proof monad wouldn't be counter productive in the long term, I think it is much better to grab the Proof.t object in coretactics.ml4 and push that info down if needed.

What? Passing the proof object to tactics is literally the **last** thing you want to do. This is breaking the logical relation between the components, just to pass a name hint. I don't even know how you would do that in practice, in addition.

I think it's safer and simpler to extend the reader component of the monad with the lemma name **for the moment**, and explicitly documenting that people shouldn't rely on that. We'll get rid of that later on, as the current priority is to make the proof state functional. I think that the proper solution is to force users to give a name to fix/cofix, and for abstract we need to design a mechanism of declarations subproofs in the kernel anyways to get rid of the side-effect hack.

**Comment by ejgallego** [comment]
> What? Passing the proof object to tactics is literally the last thing you want to do. This is breaking the logical relation between the components, just to pass a name hint. I don't even know how you would do that in practice, in addition.

Well, that's what happens in practice today; but you are right, with the current design you don't want to do that; on the other hand I am not sure how stable the current design is, it seems to me that we'll eventually refactor quite a few things including `run_tactic` which seems strange [and makes use of the proof object].

After all, the "proof object" is just a placeholder to allow vernacular-level chaining of tactics.

But note again my comment on the "proof object" living _below_ `tactics`. We may want to fix that.

> I think that the proper solution is to force users to give a name to fix/cofix, and for abstract we need to design a mechanism of declarations subproofs in the kernel anyways to get rid of the side-effect hack.

I guess that we should then implement the deprecation of the current `fix/cofix` nameless syntax and don't mess with the core layer just for this particular use case. If I am not missing anything, just by doing that #6171 should be able to go throu [taking care of abtract as an special case but IMO we can live with that].

**Review by forestjulien — APPROVED** [review]

**Comment by ppedrot** [comment]
Now that #7200 has been backported to 8.8, I think it is safe to merge this PR into master. Anybody against it?

**Review by ppedrot — APPROVED** [review]
Formal approval of the PR in its current state.

**Comment by ejgallego** [comment]
> Now that #7200 has been backported to 8.8, I think it is safe to merge this PR into master. Anybody against it?

Need to do a couple of overlays, it will take me a few days tho.

**Comment by ppedrot** [comment]
@ejgallego Do you need help for the overlays?

**Comment by ejgallego** [comment]
That's fine thanks so much, I am basically done I think.

**Comment by ppedrot** [comment]
@ejgallego up.

**Comment by ejgallego** [comment]
https://gitlab.mpi-sws.org/robbertkrebbers/coq-stdpp/issues/18

**Comment by ejgallego** [comment]
Overlays are ready, the rest is pending on these issues:

- https://github.com/fblanqui/color/issues/6
- https://github.com/math-classes/math-classes/issues/51
- https://github.com/c-corn/corn/issues/52
- https://github.com/bmsherman/topology/issues/18

**Comment by Zimmi48** [comment]
Overlays are also for backward compatible fixes. Although I can understand your reluctance to prepare the fixes yourself.

**Comment by ejgallego** [comment]
> Overlays are also for backward compatible fixes. Although I can understand your reluctance to prepare the fixes yourself.

It's not reluctance, but I think for backwards compatible fixes actually merging the fix before seems like a better process to me, as it eliminates a possible merging race.

**Comment by Zimmi48** [comment]
Oh, but I agree: backward compatible overlays should always be merged before the PR is merged. This is BTW what the CI and merging docs say. But it doesn't mean that you can't use the overlay system in the meantime to test them.

**Comment by ejgallego** [comment]
Indeed, but as this is simply adding a name I didn't feel there was something to test here.

**Review by forestjulien — APPROVED** [review]

**Comment by Zimmi48** [comment]
> - fblanqui/color#6
> - math-classes/math-classes#51
> - c-corn/corn#52
> - bmsherman/topology#18

Funny that's precisely the ones that are still in CircleCI...

**Comment by ppedrot** [comment]
I've submitted PRs for the missing developments. Let's hope they get merged quickly.

**Comment by ejgallego** [comment]
Merci Pierre-Marie

**Comment by ppedrot** [comment]
Also, what about adding a CHANGES entry? I'm not sure whether removing deprecated stuff should be mentioned there.

**Comment by ejgallego** [comment]
> Reactivity of the downstream developments we track is very unequal...

If you consider a particular CI development unreactive we can drop it, the list is not in any ways set in stone.

**Comment by ejgallego** [comment]
Note that I did a bit of refactoring in CHANGES, which looked a bit wild. Not sure if correct tho, have a look @Zimmi48 please.

**Comment by Zimmi48** [comment]
I had indeed noticed this problem and even opened #7468 about it but which had received no response (and I didn't see the point in rebasing). So I approve your refactoring, even if I would personally have kept "Tactics" and "Tactic language" separate.

**Comment by Zimmi48** [comment]
> If you consider a particular CI development unreactive we can drop it, the list is not in any ways set in stone.

Well, according to our doc, we are supposed to give a warning before actually removing it.

**Comment by ejgallego** [comment]
> Well, according to our doc, we are supposed to give a warning before actually removing it.

Sure, of course given a request to drop we would follow the procedure.

> I had indeed noticed this problem and even opened #7468 about it but which had received no response (and I didn't see the point in rebasing). So I approve your refactoring, even if I would personally have kept "Tactics" and "Tactic language" separate.

I wondered for a while about that, but I didn' see a consistent use; at the very least we should indeed try to uniformize the categories we add there.

**Pull request merged**
